### PR TITLE
Prefer with to open files

### PR DESCRIPTION
**Pull request by sam-m888**

### CODE REVIEW

**Comment by uxcq** [comment]
This is really a good idea, as Python still executes file.close() even when an exception occurs, instead of (from experience) memory issues and the file being used can be corrupted!! :+1:

**Comment by sam-m888** [comment]
@uxcq Yes, I probably should have mentioned that,  Thanks for explaining it be than I can.

File: gramps/gen/dbstate.py [hunk 1]
```diff
@@ -203,25 +203,22 @@ def lookup_family_tree(self, dbname):
...
                     else:
                         backend = "bsddb"
                     try:
                         fname = os.path.join(dirpath, "lock")
-                        ifile = open(fname, 'r', encoding='utf8')
-                        locked_by = ifile.read().strip()
-                        locked = True
+                        with open(fname, 'r', encoding='utf8') as ifile:
+                            locked_by = ifile.read().strip()
                         ifile.close()
```
Comment (QuLogic) [inline gramps/gen/dbstate.py]: You forgot to remove this `close`.

File: gramps/gen/dbstate.py [hunk 2]
```diff
@@ -203,25 +203,22 @@ def lookup_family_tree(self, dbname):
...
-                        ifile.close()
+                        with open(fname, 'r', encoding='utf8') as ifile:
+                            backend = ifile.read().strip()
                     else:
                         backend = "bsddb"
                     try:
                         fname = os.path.join(dirpath, "lock")
-                        ifile = open(fname, 'r', encoding='utf8')
-                        locked_by = ifile.read().strip()
-                        locked = True
```
Comment (QuLogic) [inline gramps/gen/dbstate.py]: This probably shouldn't be removed.

File: po/update_po.py [hunk 3]
```diff
@@ -166,23 +166,22 @@ def TipsParse(filename, mark):
...
-        #tip = tip.replace("\n<b>", "<b>") # special case tip 18
-        tip = tip.replace("</_tip>\n\n", "")
-        tip = tip.replace('"', '&quot;')
-        tips.write('char *s = N_("%s");\n' % tip)
-    tips.close()
+    with open('../data/tips.xml.in.h', 'w') as tips:
+        marklist = root.iter(mark)
+        for key in marklist:
+            tip = ElementTree.tostring(key, encoding="UTF-8", method="xml")
+            if sys.version_info[0] < 3:
```
Comment (QuLogic) [inline po/update_po.py]: This should probably be removed at some point (not necessarily in this PR.)

File: po/update_po.py [hunk 4]
```diff
@@ -262,20 +260,19 @@ def XmlParse(filename, mark):
...
-    head = open(filename + '.h', 'w')
-    
-    for key in root.iter():
-        if key.tag == '{http://www.freedesktop.org/standards/shared-mime-info}%s' % mark:
-            comment = 'char *s = N_("%s");\n' % key.text
-            head.write(comment)
-            
-    if root.tag == 'application':
+    with open(filename + '.h', 'w') as head:
+
```
Comment (QuLogic) [inline po/update_po.py]: Could drop this blank line.

File: po/update_po.py [hunk 5]
```diff
@@ -301,22 +298,19 @@ def DesktopParse(filename):
...
-    for line in lines:
-        if line[0] == '_':
-            for i in range(len(line)):
-                if line[i] == '=':
-                    val = 'char *s = N_("%s");\n' % line[i+1:len(line)]
-                    desktop.write(val)
-                    
-    desktop.close()
+    with open('../data/gramps.desktop.in.h', 'w') as desktop:
+
```
Comment (QuLogic) [inline po/update_po.py]: Could drop this blank line.

File: po/update_po.py [hunk 6]
```diff
@@ -301,22 +298,19 @@ def DesktopParse(filename):
...
+        for line in lines:
+            if line[0] == '_':
+                for i in range(len(line)):
+                    if line[i] == '=':
+                        val = 'char *s = N_("%s");\n' % line[i+1:len(line)]
+                        desktop.write(val)
+
     print ('Wrote ../data/gramps.desktop.in.h')
-
```
Comment (QuLogic) [inline po/update_po.py]: Shouldn't remove this blank line.

File: po/update_po.py [hunk 7]
```diff
@@ -342,29 +336,26 @@ def KeyParse(filename, mark):
...
-    for t in temp:
-        for i in range(len(t)):
-            if t[i] == '=':
-                val = 'char *s = N_("%s");\n' % t[i+1:len(t)]
-                key.write(val)
-    
-    key.close()
+
+    with open('../data/gramps.keys.in.h', 'w') as key:
+
```
Comment (QuLogic) [inline po/update_po.py]: Could drop this blank line.

File: po/update_po.py [hunk 8]
```diff
@@ -342,29 +336,26 @@ def KeyParse(filename, mark):
...
+        with open(filename) as f:
+            lines = [file for file in f]
```
Comment (QuLogic) [inline po/update_po.py]: `lines = f.readlines()`

File: po/update_po.py [hunk 9]
```diff
@@ -342,29 +336,26 @@ def KeyParse(filename, mark):
...
+                    temp.append(line.strip())
+
+        for t in temp:
+            for i in range(len(t)):
+                if t[i] == '=':
+                    val = 'char *s = N_("%s");\n' % t[i+1:len(t)]
+                    key.write(val)
+
     print ('Wrote ../data/gramps.keys.in.h')
-
```
Comment (QuLogic) [inline po/update_po.py]: Shouldn't remove this blank line.

File: po/update_po.py [hunk 10]
```diff
@@ -499,34 +489,30 @@ def create_filesfile():
...
-                break
-    
-    temp.close()
-    
+
+    with open('tmpfiles') as f:
+        files = [file.strip() for file in f if file and not file[0]=='#']
+
+    with open(name, 'w') as temp:
+
```
Comment (QuLogic) [inline po/update_po.py]: Could drop this blank line.

File: po/update_po.py [hunk 11]
```diff
@@ -565,14 +551,13 @@ def extract_xml():
...
-    
+
 def create_template():
     """
     Create a new file for template, if it does not exist.
     """
-    template = open('gramps.pot', 'w')
-    template.close()
-    
+    with open('gramps.pot', 'w') as template:
```
Comment (QuLogic) [inline po/update_po.py]: This context is empty; needs an indented `pass` line.

File: po/update_po.py [hunk 12]
```diff
@@ -622,21 +607,20 @@ def extract_gtkbuilder():
...
-        for line in root.iter():
-            att = line.attrib
-            if att == {'id': '0', 'translatable': 'yes'}:
-                col = 'char *s = N_("%s");\n' % line.text
-                temp.write(col)
-        root.clear()
-    
-    temp.close()
+    with open('gtklist.h', 'w') as temp:
+
```
Comment (QuLogic) [inline po/update_po.py]: Could drop this blank line.

File: gramps/gen/dbstate.py (hunk 1 above)
Comment (sam-m888) [inline gramps/gen/dbstate.py]: Ok

File: po/update_po.py (hunk 4 above)
Comment (sam-m888) [inline po/update_po.py]: Ok, Thanks

File: po/update_po.py (hunk 5 above)
Comment (sam-m888) [inline po/update_po.py]: Ok Thanks

File: po/update_po.py (hunk 7 above)
Comment (sam-m888) [inline po/update_po.py]: Ok thanks

File: po/update_po.py (hunk 10 above)
Comment (sam-m888) [inline po/update_po.py]: Ok thanks

File: po/update_po.py (hunk 11 above)
Comment (sam-m888) [inline po/update_po.py]: Good catch, thanks.

File: po/update_po.py (hunk 12 above)
Comment (sam-m888) [inline po/update_po.py]: Ok Thanks

File: gramps/gen/dbstate.py (hunk 2 above)
Comment (sam-m888) [inline gramps/gen/dbstate.py]: Which line?

Here's the original
https://github.com/gramps-project/gramps/blob/master/gramps/gen/dbstate.py#L220

File: po/update_po.py (hunk 3 above)
Comment (sam-m888) [inline po/update_po.py]: Do you mean line 173?

File: po/update_po.py (hunk 8 above)
Comment (sam-m888) [inline po/update_po.py]: Not sure what to do here?

File: gramps/gen/dbstate.py (hunk 2 above)
Comment (QuLogic) [inline gramps/gen/dbstate.py]: The `locked = True` line on which the comment is made; are you not reading these comments on GitHub?

File: po/update_po.py (hunk 3 above)
Comment (QuLogic) [inline po/update_po.py]: The entire non-Python-3 section, but it doesn't have to be now.

File: gramps/gen/dbstate.py (hunk 2 above)
Comment (sam-m888) [inline gramps/gen/dbstate.py]: Fixed

Yes, reading on Github.  More an issue of my unfamiliarity with this. Thanks for your patience. :+1:

**Comment by sam-m888** [comment]
@dsblank Is this change welcome? Does this require anything else?

**Comment by dsblank** [comment]
@sam-m888 Looks good... there were a couple of lines that @QuLogic mentioned that should not be removed. Does this look good to you, @QuLogic ?

**Comment by QuLogic** [comment]
The ones with actual content were correctly re-added. The blank lines are for PEP8; not sure how closely Gramps code follows that.

**Comment by dsblank** [comment]
PEP8: Not that close. Better to write tests :)

Thanks, all... this was a big one.

**Comment by Paul-Franklin** [comment]
"The entire non-Python-3 section, but it doesn't have to be now."

I believe that is in update_po.py (as opposed to "core" gramps).

I occasionally make a new gramps.pot on an old machine, with
only python2 installed (although I imagine Py3 could be installed).

**Comment by wroldwiedbwe** [comment]
https://github.com/gramps-project/gramps/pull/1068

**Comment by sam-m888** [comment]
more
https://github.com/gramps-project/gramps/pull/1068

**Pull request merged**